remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,562 @@
+      auto ptr2 = offset_map.find(val2);
+      if (ptr1 == offset_map.end() || ptr2 == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr1->second + ptr2->second : ptr1->second - ptr2->second);
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    }
+  }
+}
+
+// Visit a PHI node and update the offset map.
+VisitResult ForwardAliasVisitor::visitPHINode(llvm::PHINode &I) {
+  LOG(INFO) << "Entered PHI node";
+  // iterate over each operand
+  auto in_offset_map = false;
+  auto in_exclude_set = false;
+  uint64_t offset;
+  for (auto &operand : I.operands()) {
+    if (exclude.count(operand)) {
+      in_exclude_set = true;
+    } else {
+      auto ptr = offset_map.find(operand);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        if (!in_offset_map) {
+          offset = ptr->second;
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map && in_exclude_set) {
+    // fail if some operands are excluded and others are state offsets
+    return VisitResult::Error;
+  } else if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else if (in_exclude_set) {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void RemoveDeadStores(const std::unordered_set<llvm::Instruction *> &dead_stores) {
+  for (auto *store : dead_stores) {
+    store->eraseFromParent();
+  }
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(const ScopeMap &scope_to_offset_, const llvm::FunctionType *lft_)
+  : scope_to_offset(scope_to_offset_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove(),
+    lft(lft_) {}
+
+// Get every terminating basic block from the function `func`.
+void LiveSetBlockVisitor::AddFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry set
+      block_map.insert({&block, LiveSet()});
+    }
+  }
+}
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  bool progress = true;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty() && progress) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    progress = false;
```

Comment:
I don't think you need a `progress` tracker or two separate work lists. You should be able to just have a single vector as a worklist, and alway add on to the end of it.

Reply:
Couldn't that cause iterator invalidation to add to a vector worklist while iterating it, [in the case that the capacity changes](http://kera.name/articles/2011/06/iterator-invalidation-rules-c0x/)?
The `progress` tracker was added because the program would loop infinitely without it before, so I'll remove it.